cabal-install/Distribution/Client/HttpUtils.hs

```diff
@@ -147,16 +147,30 @@ downloadURI verbosity uri path = do
     Left err   -> die $ "Failed to download " ++ show uri ++ " : " ++ show err
     Right rsp -> case rspCode rsp of
       (2,0,0) -> do
-        info verbosity ("Downloaded to " ++ path)
-        writeFileAtomic path $ rspBody rsp
-        return (FileDownloaded path)
+        let index = rspBody rsp
+        case lookupHeader HdrContentLength (rspHeaders rsp) >>= readMaybe of
+            Just contentLength -> do
+              let indexLength = ByteString.length index
+              when (indexLength /= contentLength) $
+                  die $ "Failed to download " ++ show uri
+                        ++ " : Expected content length "   ++ (show contentLength)
+                        ++ " doesn't match actual length " ++ (show indexLength)
+
+              info verbosity ("Downloaded to " ++ path)
+              writeFileAtomic path index 
+              return (FileDownloaded path)
+            Nothing -> die $ "Failed to download " ++ show uri ++ " cannot verify index file length"
       (3,0,4) -> do
         notice verbosity "Skipping download: Local and remote files match."
         return FileAlreadyInCache
       (_,_,_) -> return (FileDownloaded path)
-      --FIXME: check the content-length header matches the body length.
       --TODO: stream the download into the file rather than buffering the whole
       --      thing in memory.
+  where
+    readMaybe :: Read a => String -> Maybe a
```

Comment:
I have added it to D.C.Utils because in the cabal library there is no use of it.